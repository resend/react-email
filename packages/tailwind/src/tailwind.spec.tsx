import { Button } from '@react-email/button';
import { Head } from '@react-email/head';
import { Heading } from '@react-email/heading';
/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-explicit-any */
import { Hr } from '@react-email/hr';
import { Html } from '@react-email/html';
import { Link } from '@react-email/link';
import { render } from '@react-email/render';
import { ResponsiveColumn, ResponsiveRow } from '@responsive-email/react-email';
import React from 'react';
import { Tailwind } from '.';
import type { TailwindConfig } from '.';

describe('Tailwind component', () => {
  it('should allow for complex children manipulation', async () => {
    const actualOutput = await render(
      <Tailwind>
        <ResponsiveRow>
          <ResponsiveColumn>This is the first column</ResponsiveColumn>
          <ResponsiveColumn>This is the second column</ResponsiveColumn>
        </ResponsiveRow>
      </Tailwind>,
    );
    expect(actualOutput).toMatchSnapshot();
  });

  it('should work with class manipulation done on components', async () => {
    const MyComponnt = (props: {
      className?: string;
      style?: React.CSSProperties;
    }) => {
      expect(
        props.style,
        'Styles should be generated the same for a component',
      ).toEqual({
        color: 'rgb(96,165,250)',
        padding: '1rem',
      });
      return (
        <div
          className={`${props.className} bg-red-500`}
          style={{ ...props.style, padding: '4px' }}
        />
      );
    };

    expect(
      await render(
        <Tailwind>
          <MyComponnt className="text-blue-400 p-4" />
        </Tailwind>,
      ),
    ).toMatchSnapshot();
  });

  it("should work properly with 'no-underline'", async () => {
    const actualOutput = await render(
      <Html>
        <body>
          <Tailwind>
            <p className="text-black text-[14px] leading-[24px]">
              or copy and paste this URL into your browser:{' '}
              <Link
                className="text-blue-600 no-underline other"
                href="https://react.email"
              >
                https://react.email
              </Link>
            </p>
            <p className="text-black text-[14px] leading-[24px]">
              or copy and paste this URL into your browser:{' '}
              <Link
                className="text-blue-600 no-underline"
                href="https://react.email"
              >
                https://react.email
              </Link>
            </p>
          </Tailwind>
        </body>
      </Html>,
    );
    expect(actualOutput).toMatchSnapshot();
  });

  describe('Inline styles', () => {
    it('should render children with inline Tailwind styles', async () => {
      const actualOutput = await render(
        <Tailwind>
          <div className="bg-white" />
        </Tailwind>,
      );

      expect(actualOutput).not.toBeNull();
    });
  });

  // test("with React context and custom components", () => {
  //   const SharedDataContext = React.createContext<{ name: string } | undefined>(undefined);
  //
  //   const IsGreat = () => {
  //     // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  //     const sharedData = React.useContext(SharedDataContext)!;
  //
  //     expect(sharedData).toBeDefined();
  //
  //     return <p className="text-red-500 sm:text-blue-300">
  //       {sharedData.name} is great!
  //     </p>;
  //   };
  //
  //   render(
  //     <Html>
  //       <Tailwind>
  //         <Head/>
  //         <SharedDataContext.Provider value={{ name: 'React Email' }}>
  //           <body className="bg-slate-900 text-gray-200">
  //             <IsGreat/>
  //           </body>
  //         </SharedDataContext.Provider>
  //       </Tailwind>
  //     </Html>
  //   );
  // });

  test('<Button className="px-3 py-2 mt-8 text-sm text-gray-200 bg-blue-600 rounded-md">', async () => {
    const actualOutput = await render(
      <Tailwind>
        <Button className="px-3 py-2 mt-8 text-sm text-gray-200 bg-blue-600 rounded-md">
          Testing button
        </Button>
        Testing
      </Tailwind>,
    );

    expect(actualOutput).toMatchSnapshot();
  });

  it('should work with custom components with fragment at the root', async () => {
    const Wrapper = (props: { children: React.ReactNode }) => {
      return <Tailwind>{props.children}</Tailwind>;
    };

    const Brand = () => {
      return (
        <>
          <div className="p-[20px]">
            <p className="font-bold text-[50px]">React Email</p>
          </div>
          <div className="p-[20px]">
            <p className="font-bold text-[50px]">React Email</p>
          </div>
        </>
      );
    };

    const EmailTemplate = () => {
      return (
        <Wrapper>
          <div className="text-[50px] leading-[1] mt-[100px]">Hello world</div>
          <Brand />
        </Wrapper>
      );
    };

    const actualOutput = await render(EmailTemplate());

    expect(actualOutput).toMatchSnapshot();
  });

  test('it should not generate styles from text', async () => {
    expect(
      await render(<Tailwind>container bg-red-500 bg-blue-300</Tailwind>),
    ).toMatchSnapshot();
  });

  it('should work with components that return children', async () => {
    const Wrapper = (props: { children: React.ReactNode }) => {
      return <Tailwind>{props.children}</Tailwind>;
    };

    const Brand = () => {
      return (
        <div className="p-[20px]">
          <p className="font-bold text-[50px]">React Email</p>
        </div>
      );
    };

    const EmailTemplate = () => {
      return (
        <Wrapper>
          <div className="text-[50px] leading-[1] mt-[100px]">Hello world</div>
          <Brand />
        </Wrapper>
      );
    };

    const actualOutput = await render(EmailTemplate());

    expect(actualOutput).toMatchSnapshot();
  });

  it('should work with Heading component', async () => {
    const EmailTemplate = () => {
      return (
        <Tailwind>
          Hello
          <Heading>My testing heading</Heading>
          friends
        </Tailwind>
      );
    };

    expect(await render(<EmailTemplate />)).toMatchSnapshot();
  });

  it('should work with components that use React.forwardRef', async () => {
    const Wrapper = (props: { children: React.ReactNode }) => {
      return <Tailwind>{props.children}</Tailwind>;
    };

    const Brand = React.forwardRef<HTMLDivElement>((ref, props) => {
      return (
        <div
          className="p-[20px]"
          ref={ref as React.LegacyRef<HTMLDivElement>}
          {...props}
        >
          <p className="font-bold text-[50px]">React Email</p>
        </div>
      );
    });
    Brand.displayName = 'Brand';

    const EmailTemplate = () => {
      return (
        <Wrapper>
          <div className="text-[50px] leading-[1] mt-[100px]">Hello world</div>
          <Brand />
        </Wrapper>
      );
    };

    const actualOutput = await render(EmailTemplate());

    expect(actualOutput).toMatchSnapshot();
  });

  it('should be able to use background image', async () => {
    const actualOutput = await render(
      <Tailwind>
        <div className="bg-[url(https://example.com/image.png)]" />
      </Tailwind>,
    );

    expect(actualOutput).toMatchSnapshot();
  });

  it('should not override inline styles with Tailwind styles', async () => {
    const actualOutput = await render(
      <Tailwind>
        <div
          className="bg-black text-[16px]"
          style={{ backgroundColor: 'red', fontSize: '12px' }}
        />
      </Tailwind>,
    );

    expect(actualOutput).toMatchSnapshot();
  });

  it('should override component styles with Tailwind styles', async () => {
    const actualOutput = await render(
      <Tailwind>
        <Hr className="w-12" />
      </Tailwind>,
    );

    expect(actualOutput).toContain('width:3rem');
  });

  it('should preserve mso styles', async () => {
    const actualOutput = await render(
      <Html>
        <Tailwind>
          <Head />
          <span
            dangerouslySetInnerHTML={{
              __html: `<!--[if mso]><i style="letter-spacing: 10px;mso-font-width:-100%;" hidden>&nbsp;</i><![endif]-->`,
            }}
          />
          <div className="bg-white sm:bg-red-50 sm:text-sm md:text-lg custom-class" />
        </Tailwind>
      </Html>,
    );

    expect(actualOutput).toMatchSnapshot();
  });

  it('should recognize custom responsive screen', async () => {
    const config: TailwindConfig = {
      theme: {
        screens: {
          sm: { min: '640px' },
          md: { min: '768px' },
          lg: { min: '1024px' },
          xl: { min: '1280px' },
          '2xl': { min: '1536px' },
        },
      },
    };
    const actualOutput = await render(
      <Html>
        <Tailwind config={config}>
          <Head />
          <div className="bg-red-100 xl:bg-green-500">Test</div>
          <div className="2xl:bg-blue-500">Test</div>
        </Tailwind>
      </Html>,
    );

    expect(actualOutput).toMatchSnapshot();
  });

  it('should work with calc() with + sign', async () => {
    const actualOutput = await render(
      <Tailwind>
        <head />
        <div className="max-h-[calc(50px+3rem)] lg:max-h-[calc(50px+5rem)] bg-red-100">
          <div className="h-[200px]">something tall</div>
        </div>
      </Tailwind>,
    );

    expect(actualOutput).toMatchSnapshot();
  });
});

describe('Responsive styles', () => {
  /*
    This test is because of https://github.com/resend/react-email/issues/1112
    which was being caused because we required to, either have our <Head> component,
    or a <head> element directly inside the <Tailwind> component for media queries to be applied
    onto. The problem with this approach was that the check to see if an element was an instance of
    the <Head> component fails after minification as we did it by the function name.

    The best solution is to check for the Head element on arbitrarily deep levels of the React tree
    and apply the styles there. This also fixes the issue where it would not be allowed to use
    Tailwind classes on the <html> element as the <head> would be required directly bellow Tailwind.
  */
  it('should work with arbitrarily deep (in the React tree) <head> elements', async () => {
    expect(
      await render(
        <Tailwind>
          <html lang="en">
            <head />
            <body>
              <div className="bg-red-200 sm:bg-red-300 md:bg-red-400 lg:bg-red-500" />
            </body>
          </html>
        </Tailwind>,
      ),
    ).toMatchSnapshot();

    const MyHead = (props: Record<string, any>) => {
      return <head {...props} />;
    };

    expect(
      await render(
        <Tailwind>
          <html lang="en">
            <MyHead />
            <body>
              <div className="bg-red-200 sm:bg-red-300 md:bg-red-400 lg:bg-red-500" />
            </body>
          </html>
        </Tailwind>,
      ),
    ).toMatchSnapshot();
  });

  it('should not have duplicate media queries', async () => {
    const Body = (props: { className: string; children: React.ReactNode }) => {
      return <body className={props.className}>{props.children}</body>;
    };
    const output = await render(
      <Tailwind>
        <Head />
        <Body className="bg-white my-auto mx-auto font-sans md:px-[64px]">
          <div className="md:px-[64px]" />
        </Body>
      </Tailwind>,
    );

    expect(output).toMatchSnapshot();
  });

  it('should add css to <head/> and keep responsive class names', async () => {
    const actualOutput = await render(
      <html lang="en">
        <Tailwind>
          <head />
          <body>
            <div className="bg-red-200 sm:bg-red-300 md:bg-red-400 lg:bg-red-500" />
          </body>
        </Tailwind>
      </html>,
    );

    expect(actualOutput).toMatchSnapshot();
  });

  it('should throw error when used without the head and with media query class names only very deeply nested', async () => {
    const Component1 = (props: Record<string, any>) => {
      return (
        <div {...props} className="w-40 h-30 sm:h-10 sm:w-10">
          {props.children}
        </div>
      );
    };
    const Component2 = (props: Record<string, any>) => {
      return (
        <div {...props}>
          <Component1>{props.children}</Component1>
        </div>
      );
    };
    const Component3 = (props: Record<string, any>) => {
      return (
        <div {...props}>
          <Component2>{props.children}</Component2>
        </div>
      );
    };

    function renderComplexEmailWithoutHead() {
      return render(
        <Tailwind>
          <div className="bg-red-300">
            <Component3 className="random-classname w-full">
              <div className="w-50">Testing</div>
            </Component3>
          </div>
        </Tailwind>,
      );
    }

    await expect(
      renderComplexEmailWithoutHead,
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it('should work with relatively complex media query utilities', async () => {
    const Email = () => {
      return (
        <Tailwind>
          <Head />
          <p className="text-blue-700 max-sm:text-red-600">I am some text</p>
        </Tailwind>
      );
    };

    expect(await render(<Email />)).toMatchSnapshot();
  });

  it('should throw an error when used without a <head/>', async () => {
    function noHead() {
      return render(
        <Tailwind>
          <html lang="en">
            {/* <Head></Head> */}
            <div className="bg-red-200 sm:bg-red-500" />
          </html>
        </Tailwind>,
      );
    }
    await expect(noHead).rejects.toThrowErrorMatchingSnapshot();
  });

  it('should persist existing <head/> elements', async () => {
    const actualOutput = await render(
      <html lang="en">
        <Tailwind>
          <head>
            <style />
            <link />
          </head>
          <body>
            <div className="bg-red-200 sm:bg-red-500" />
          </body>
        </Tailwind>
      </html>,
    );

    expect(actualOutput).toMatchSnapshot();
  });
});

describe('Custom theme config', () => {
  it('should be able to use custom colors', async () => {
    const config: TailwindConfig = {
      theme: {
        extend: {
          colors: {
            custom: '#1fb6ff',
          },
        },
      },
    };

    const actualOutput = await render(
      <Tailwind config={config}>
        <div className="text-custom bg-custom" />
      </Tailwind>,
    );

    expect(actualOutput).toMatchSnapshot();
  });

  it('should be able to use custom fonts', async () => {
    const config: TailwindConfig = {
      theme: {
        extend: {
          fontFamily: {
            sans: ['Graphik', 'sans-serif'],
            serif: ['Merriweather', 'serif'],
          },
        },
      },
    };

    const actualOutput = await render(
      <Tailwind config={config}>
        <div className="font-sans" />
        <div className="font-serif" />
      </Tailwind>,
    );

    expect(actualOutput).toMatchSnapshot();
  });

  it('should be able to use custom spacing', async () => {
    const config: TailwindConfig = {
      theme: {
        extend: {
          spacing: {
            '8xl': '96rem',
          },
        },
      },
    };
    const actualOutput = await render(
      <Tailwind config={config}>
        <div className="m-8xl" />
      </Tailwind>,
    );
    expect(actualOutput).toMatchSnapshot();
  });

  it('should be able to use custom border radius', async () => {
    const config: TailwindConfig = {
      theme: {
        extend: {
          borderRadius: {
            '4xl': '2rem',
          },
        },
      },
    };
    const actualOutput = await render(
      <Tailwind config={config}>
        <div className="rounded-4xl" />
      </Tailwind>,
    );
    expect(actualOutput).toMatchSnapshot();
  });

  it('should be able to use custom text alignment', async () => {
    const config: TailwindConfig = {
      theme: {
        extend: {
          textAlign: {
            justify: 'justify',
          },
        },
      },
    };

    const actualOutput = await render(
      <Tailwind config={config}>
        <div className="text-justify" />
      </Tailwind>,
    );

    expect(actualOutput).toMatchSnapshot();
  });
});

describe('Custom plugins config', () => {
  it('should be able to use custom plugins', async () => {
    const config: TailwindConfig = {
      plugins: [
        ({ addUtilities }: any) => {
          const newUtilities = {
            '.border-custom': {
              border: '2px solid',
            },
          };

          addUtilities(newUtilities);
        },
      ],
    };

    const actualOutput = await render(
      <Tailwind config={config}>
        <div className="border-custom" />
      </Tailwind>,
    );

    expect(actualOutput).toMatchSnapshot();
  });

  it('should be able to use custom plugins with responsive styles', async () => {
    const config: TailwindConfig = {
      plugins: [
        ({ addUtilities }: any) => {
          const newUtilities = {
            '.border-custom': {
              border: '2px solid',
            },
          };

          addUtilities(newUtilities);
        },
      ],
    };

    const actualOutput = await render(
      <html lang="en">
        <Tailwind config={config}>
          <head />
          <body>
            <div className="border-custom sm:border-custom" />
          </body>
        </Tailwind>
      </html>,
    );

    expect(actualOutput).toMatchSnapshot();
  });
});
