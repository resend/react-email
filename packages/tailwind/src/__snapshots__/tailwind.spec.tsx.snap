// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tailwind component > <Button className="px-3 py-2 mt-8 text-sm text-gray-200 bg-blue-600 rounded-md"> 1`] = `"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"><!--$--><a style="margin-top:2rem;border-radius:0.375rem;background-color:rgb(21,93,252);color:rgb(229,231,235);font-size:0.875rem;line-height:100%;text-decoration:none;display:inline-block;max-width:100%;mso-padding-alt:0px" target="_blank"><span><!--[if mso]><i style="mso-font-width:0%;mso-text-raise:0" hidden></i><![endif]--></span><span style="max-width:100%;display:inline-block;line-height:120%;mso-padding-alt:0px">Testing button</span><span><!--[if mso]><i style="mso-font-width:0%" hidden>&#8203;</i><![endif]--></span></a>Testing<!--/$-->"`;

exports[`Tailwind component > it should not generate styles from text 1`] = `"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"><!--$-->container bg-red-500 bg-blue-300<!--/$-->"`;

exports[`Tailwind component > should allow for complex children manipulation 1`] = `"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"><!--$--><table align="center" width="100%" border="0" cellPadding="0" cellSpacing="0" role="presentation" style="text-align:center;font-size:0"><tbody><tr><td style="padding:0px 0px 0px 0px"><table align="center" width="100%" border="0" cellPadding="0" cellSpacing="0" role="presentation" style="max-width:300px;display:inline-block;vertical-align:top;font-size:16px;box-sizing:border-box"><tbody><tr><td>This is the first column</td></tr></tbody></table><table align="center" width="100%" border="0" cellPadding="0" cellSpacing="0" role="presentation" style="max-width:300px;display:inline-block;vertical-align:top;font-size:16px;box-sizing:border-box"><tbody><tr><td>This is the second column</td></tr></tbody></table></td></tr></tbody></table><!--/$-->"`;

exports[`Tailwind component > should be able to use background image 1`] = `"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"><!--$--><div style="background-image:url(https://example.com/image.png)"></div><!--/$-->"`;

exports[`Tailwind component > should not override inline styles with Tailwind styles 1`] = `"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"><!--$--><div style="background-color:red;font-size:12px"></div><!--/$-->"`;

exports[`Tailwind component > should override component styles with Tailwind styles 1`] = `"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"><!--$--><hr style="width:100%;border:none;border-top:1px solid #eaeaea"/><!--/$-->"`;

exports[`Tailwind component > should preserve mso styles 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html dir="ltr" lang="en">
  <head>
    <meta content="text/html; charset=UTF-8" http-equiv="Content-Type" />
    <meta name="x-apple-disable-message-reformatting" />
    <!--$-->
    <style>
      .sm_bg-red-50{@media (width>=40rem){background-color:rgb(254,242,242)!important}}.sm_text-sm{@media (width>=40rem){font-size:0.875rem!important;line-height:1.4285714285714286!important}}.md_text-lg{@media (width>=48rem){font-size:1.125rem!important;line-height:1.5555555555555556!important}}
    </style></head
  ><span
    ><!--[if mso]><i style="letter-spacing: 10px;mso-font-width:-100%;" hidden>&nbsp;</i><![endif]--></span
  >
  <div
    class="custom-class sm_bg-red-50 sm_text-sm md_text-lg"
    style="background-color:rgb(255,255,255)"></div>
  <!--/$-->
</html>
"
`;

exports[`Tailwind component > should recognize custom responsive screen 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html dir="ltr" lang="en">
  <head>
    <meta content="text/html; charset=UTF-8" http-equiv="Content-Type" />
    <meta name="x-apple-disable-message-reformatting" />
    <!--$-->
    <style>
      .xl_bg-green-500{@media (width>=1280px){background-color:rgb(0,201,80)!important}}.twoxl_bg-blue-500{@media (width>=1536px){background-color:rgb(43,127,255)!important}}
    </style>
  </head>
  <div class="xl_bg-green-500" style="background-color:rgb(255,226,226)">
    Test
  </div>
  <div class="twoxl_bg-blue-500">Test</div>
  <!--/$-->
</html>
"
`;

exports[`Tailwind component > should render children with inline Tailwind styles 1`] = `"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"><!--$--><div style="background-color:rgb(255,255,255)"></div><!--/$-->"`;

exports[`Tailwind component > should work properly with 'no-underline' 1`] = `"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"><html dir="ltr" lang="en"><head></head><!--$--><body><p style="font-size:14px;color:rgb(0,0,0);line-height:24px">or copy and paste this URL into your browser:<!-- --> <a class="other" href="https://react.email" style="color:#067df7;text-decoration-line:none" target="_blank">https://react.email</a></p><p style="font-size:14px;color:rgb(0,0,0);line-height:24px">or copy and paste this URL into your browser:<!-- --> <a href="https://react.email" style="color:#067df7;text-decoration-line:none" target="_blank">https://react.email</a></p><!--/$--></body></html>"`;

exports[`Tailwind component > should work with Heading component 1`] = `"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"><!--$-->Hello<h1>My testing heading</h1>friends<!--/$-->"`;

exports[`Tailwind component > should work with blocklist 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<head>
  <meta content="text/html; charset=UTF-8" http-equiv="Content-Type" />
  <meta name="x-apple-disable-message-reformatting" />
  <!--$-->
  <style>
    .md_p-4{@media (width>=48rem){padding:1rem!important}}
  </style>
</head>
<body>
  <button type="button" class="bg-blue-600 md_p-4">Click me</button
  ><!--/$-->
</body>
"
`;

exports[`Tailwind component > should work with calc() with + sign 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<head>
  <!--$-->
  <style>
    .lg_max-h-calc50pxplus5rem{@media (width>=64rem){max-height:calc(50px + 5rem)!important}}
  </style>
</head>
<div
  class="lg_max-h-calc50pxplus5rem"
  style="max-height:calc(50px + 3rem);background-color:rgb(255,226,226)">
  <div style="height:200px">something tall</div>
</div>
<!--/$-->
"
`;

exports[`Tailwind component > should work with class manipulation done on components 1`] = `"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"><!--$--><div style="padding:4px;color:rgb(81,162,255);background-color:rgb(251,44,54)"></div><!--/$-->"`;

exports[`Tailwind component > should work with components that return children 1`] = `"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"><!--$--><div style="margin-top:100px;font-size:50px;line-height:1">Hello world</div><div style="padding:20px"><p style="font-weight:700;font-size:50px">React Email</p></div><!--/$-->"`;

exports[`Tailwind component > should work with components that use React.forwardRef 1`] = `"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"><!--$--><div style="margin-top:100px;font-size:50px;line-height:1">Hello world</div><div style="padding:20px"><p style="font-weight:700;font-size:50px">React Email</p></div><!--/$-->"`;

exports[`Tailwind component > should work with custom components with fragment at the root 1`] = `"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"><!--$--><div style="margin-top:100px;font-size:50px;line-height:1">Hello world</div><div style="padding:20px"><p style="font-weight:700;font-size:50px">React Email</p></div><div style="padding:20px"><p style="font-weight:700;font-size:50px">React Email</p></div><!--/$-->"`;

exports[`Tailwind component > with custom plugins config > should be able to use custom plugins 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<!--$-->
<div style="border:2px solid"></div>
<!--/$-->
"
`;

exports[`Tailwind component > with custom plugins config > should be able to use custom plugins with responsive styles 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html lang="en">
  <head>
    <!--$-->
    <style>
      .sm_border-custom{@media (width>=40rem){border:2px solid!important}}
    </style>
  </head>
  <body>
    <div class="sm_border-custom" style="border:2px solid"></div>
    <!--/$-->
  </body>
</html>
"
`;

exports[`Tailwind component > with custom theme config > should be able to use custom border radius 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<!--$-->
<div style="border-radius:2rem"></div>
<!--/$-->
"
`;

exports[`Tailwind component > with custom theme config > should be able to use custom colors 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<!--$-->
<div style="background-color:rgb(31,182,255);color:rgb(31,182,255)"></div>
<!--/$-->
"
`;

exports[`Tailwind component > with custom theme config > should be able to use custom fonts 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<!--$-->
<div style="font-family:Graphik,sans-serif"></div>
<div style="font-family:Merriweather,serif"></div>
<!--/$-->
"
`;

exports[`Tailwind component > with custom theme config > should be able to use custom spacing 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<!--$-->
<div style="margin:96rem"></div>
<!--/$-->
"
`;

exports[`Tailwind component > with custom theme config > should be able to use custom text alignment 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<!--$-->
<div style="text-align:justify"></div>
<!--/$-->
"
`;

exports[`Tailwind component > with non-inlinable styles > should add css to <head/> and keep class names 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html lang="en">
  <head>
    <!--$-->
    <style>
      .hover_bg-red-600{&:hover{@media (hover:hover){background-color:rgb(231,0,11)!important}}}.focus_bg-red-700{&:focus{background-color:rgb(193,0,7)!important}}.sm_bg-red-300{@media (width>=40rem){background-color:rgb(255,162,162)!important}}.sm_hover_bg-red-200{@media (width>=40rem){&:hover{@media (hover:hover){background-color:rgb(255,201,201)!important}}}}.md_bg-red-400{@media (width>=48rem){background-color:rgb(255,100,103)!important}}.lg_bg-red-500{@media (width>=64rem){background-color:rgb(251,44,54)!important}}
    </style>
  </head>
  <body>
    <div
      class="hover_bg-red-600 focus_bg-red-700 sm_bg-red-300 sm_hover_bg-red-200 md_bg-red-400 lg_bg-red-500"
      style="background-color:rgb(255,201,201)"></div>
    <!--/$-->
  </body>
</html>
"
`;

exports[`Tailwind component > with non-inlinable styles > should persist existing <head/> elements 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html lang="en">
  <head>
    <!--$-->
    <style>
      .sm_bg-red-500{@media (width>=40rem){background-color:rgb(251,44,54)!important}}
    </style>
    <style></style>
    <link />
  </head>
  <body>
    <div class="sm_bg-red-500" style="background-color:rgb(255,201,201)"></div>
    <!--/$-->
  </body>
</html>
"
`;

exports[`Tailwind component > with non-inlinable styles > should throw an error when used without a <head/> 1`] = `
[Error: You are trying to use the following Tailwind classes that cannot be inlined: sm:bg-red-500.
For the media queries to work properly on rendering, they need to be added into a <style> tag inside of a <head> tag,
the Tailwind component tried finding a <head> element but just wasn't able to find it.

Make sure that you have a <head> element at some point inside of the <Tailwind> component at any depth. 
This can also be our <Head> component.

If you do already have a <head> element at some depth, 
please file a bug https://github.com/resend/react-email/issues/new?assignees=&labels=Type%3A+Bug&projects=&template=1.bug_report.yml.]
`;

exports[`Tailwind component > with non-inlinable styles > should throw error when used without the head and with media query class names very deeply nested 1`] = `
[Error: You are trying to use the following Tailwind classes that cannot be inlined: sm:h-10 sm:w-10.
For the media queries to work properly on rendering, they need to be added into a <style> tag inside of a <head> tag,
the Tailwind component tried finding a <head> element but just wasn't able to find it.

Make sure that you have a <head> element at some point inside of the <Tailwind> component at any depth. 
This can also be our <Head> component.

If you do already have a <head> element at some depth, 
please file a bug https://github.com/resend/react-email/issues/new?assignees=&labels=Type%3A+Bug&projects=&template=1.bug_report.yml.]
`;

exports[`Tailwind component > with non-inlinable styles > should work with arbitrarily deep (in the React tree) <head> elements 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html lang="en">
  <head>
    <!--$-->
    <style>
      .sm_bg-red-300{@media (width>=40rem){background-color:rgb(255,162,162)!important}}.md_bg-red-400{@media (width>=48rem){background-color:rgb(255,100,103)!important}}.lg_bg-red-500{@media (width>=64rem){background-color:rgb(251,44,54)!important}}
    </style>
  </head>
  <body>
    <div
      class="sm_bg-red-300 md_bg-red-400 lg_bg-red-500"
      style="background-color:rgb(255,201,201)"></div>
    <!--/$-->
  </body>
</html>
"
`;

exports[`Tailwind component > with non-inlinable styles > should work with arbitrarily deep (in the React tree) <head> elements 2`] = `"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd"><html lang="en"><head><!--$--><style>.sm_bg-red-300{@media (width>=40rem){background-color:rgb(255,162,162)!important}}.md_bg-red-400{@media (width>=48rem){background-color:rgb(255,100,103)!important}}.lg_bg-red-500{@media (width>=64rem){background-color:rgb(251,44,54)!important}}</style></head><body><div class="sm_bg-red-300 md_bg-red-400 lg_bg-red-500" style="background-color:rgb(255,201,201)"></div><!--/$--></body></html>"`;

exports[`Tailwind component > with non-inlinable styles > should work with relatively complex media query utilities 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<head>
  <meta content="text/html; charset=UTF-8" http-equiv="Content-Type" />
  <meta name="x-apple-disable-message-reformatting" />
  <!--$-->
  <style>
    .max-sm_text-red-600{@media (width<40rem){color:rgb(231,0,11)!important}}
  </style>
</head>
<p class="max-sm_text-red-600" style="color:rgb(20,71,230)">I am some text</p>
<!--/$-->
"
`;
